src/server/scripts/Northrend/Naxxramas/boss_thaddius.cpp

```diff
@@ -110,168 +155,302 @@ class boss_thaddius : public CreatureScript
-                case ACTION_FEUGEN_DIED:
-                    checkFeugenAlive = false;
-                    break;
-                case ACTION_STALAGG_RESET:
-                    checkStalaggAlive = true;
-                    break;
-                case ACTION_STALAGG_DIED:
-                    checkStalaggAlive = false;
-                    break;
+                _JustDied();
+                me->setActive(false);
+                if (Creature* stalagg = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_STALAGG)))
+                    stalagg->setActive(false);
+                if (Creature* feugen = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_FEUGEN)))
+                    feugen->setActive(false);
+                Talk(SAY_DEATH);
             }
 
-            if (!checkFeugenAlive && !checkStalaggAlive)
+            void DoAction(int32 action) override
             {
-                me->RemoveFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_IMMUNE_TO_PC | UNIT_FLAG_NOT_SELECTABLE | UNIT_FLAG_STUNNED);
-                // REACT_AGGRESSIVE only reset when he takes damage.
-                DoZoneInCombat();
+                switch (action)
+                {
+                    case ACTION_RESET_ENCOUNTER_TIMER:
+                        if (_phase == PHASE_RESETTING)
+                            ResetEncounter();
+                        break;
+                    case ACTION_FEUGEN_RESET:
+                    case ACTION_STALAGG_RESET:
+                        if (_phase != PHASE_NOT_ENGAGED && _phase != PHASE_RESETTING)
+                            BeginResetEncounter();
+                        break;
+                    case ACTION_FEUGEN_AGGRO:
+                    case ACTION_STALAGG_AGGRO:
+                        if (_phase == PHASE_RESETTING)
+                            return BeginResetEncounter();
+                        if (_phase != PHASE_NOT_ENGAGED)
+                            return;
+                        _phase = PHASE_PETS;
+                        events.SetPhase(PHASE_PETS);
+
+                        shockingEligibility = true;
+                        
+                        if (!instance->CheckRequiredBosses(BOSS_THADDIUS))
+                            return BeginResetEncounter();
+                        instance->SetBossState(BOSS_THADDIUS, IN_PROGRESS);
+
+                        me->setActive(true);
+                        if (Creature* stalagg = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_STALAGG)))
+                            stalagg->setActive(true);
+                        if (Creature* feugen = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_FEUGEN)))
+                            feugen->setActive(true);
+                        break;
+                    case ACTION_FEUGEN_DIED:
+                        if (Creature* feugen = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_FEUGEN)))
+                            if (feugen->AI())
```

Comment:
> Wouldn't that cause a crash/UB if someone fiddles with the creature_template and removes the AI for feugen/stalagg (but leaves it for thaddius) or vice versa?

If someone fiddles with something they shouldn't we want it to crash so that the errors/typos can be corrected. The code should be robust indeed but we don't want it to hide problems.